backend/src/lib.rs

```diff
@@ -1,104 +1,186 @@
+        Err(Error::NotApplicable)
+    }
 }
 
-pub trait ProverAggregationWithParams {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof;
-}
+/// Concrete dynamic dispatch Backend object, for backends without setup.
+struct ConcreteBackendWithoutSetup<B>(B);
 
-#[cfg(feature = "halo2")]
-pub struct Halo2Backend;
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F>> Backend<F> for ConcreteBackendWithoutSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        self.0.prove(pil, fixed, witness, prev_proof, output_dir)
+    }
 
-#[cfg(feature = "halo2")]
-pub struct Halo2MockBackend;
+    fn write_setup(&self, _output: &mut dyn io::Write) -> Result<(), Error> {
+        Err(Error::NotApplicable)
+    }
+}
 
-#[cfg(feature = "halo2")]
-pub struct Halo2AggregationBackend;
+/// Factory for backends with setup.
+struct WithSetupFactory<B>(PhantomData<B>);
 
-#[cfg(feature = "halo2")]
-impl ProverWithParams for Halo2Backend {
-    fn prove<T: FieldElement, R: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof> {
-        Some(halo2::prove_ast_read_params(pil, fixed, witness, params))
+/// Factory implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F> + 'static> BackendFactory<F> for WithSetupFactory<B> {
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
+        Box::new(ConcreteBackendWithSetup(B::new_setup_from_params(degree)))
     }
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params {
-        halo2::generate_params::<T>(size)
+    fn load_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
+        Ok(Box::new(ConcreteBackendWithSetup(B::load_setup(input)?)))
     }
 }
 
-#[cfg(feature = "halo2")]
-impl ProverWithoutParams for Halo2MockBackend {
-    fn prove<T: FieldElement>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-    ) -> Option<Proof> {
-        halo2::mock_prove(pil, fixed, witness);
-        None
+/// Concrete dynamic dispatch Backend object, for backends with setup.
+struct ConcreteBackendWithSetup<B>(B);
+
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F>> Backend<F> for ConcreteBackendWithSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
```

Comment:
witness option why

Reply:
Because Pilcom backend can write the json file even without witness. Don't know how to fix.

Reply:
We don't have such a use case at the moment. All our use cases require a witness, so I think we should still enforce it

Reply:
removed option

Reply:
test failed because of it